### PR TITLE
Implement special case handling of EnvoyFilter for cluster having transport_socket_matches defined

### PR DESCRIPTION
**Pull request by JuniorJPDJ**
**Please provide a description of this PR:**
Provides workaround mentioned in #42257 and would allow to implement #44342 without explict support from istio - using EnvoyFilter.

Allows modifying transport_sockets in transport_socket_matches array when type is going to be changed, allows matches overwrite for transport_socket_matches.
Matches array elements by transport_socket_matches.name

### CODE REVIEW

**Review by hzxuzhonghu — COMMENTED** [review]
please add some tests for it

**Comment by JuniorJPDJ** [comment]
I'll try, but I'm not golang programmer ;/

I just built patched 1.17.2 version and transport_socket upstream proxy protocol EnvoyFilter works!

image

```
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: proxy-protocol
  namespace: istio-system
spec:
  configPatches:
  - applyTo: CLUSTER
    match:
      context: GATEWAY
      cluster:
        portNumber: 22
        service: big-boy.istio-system.svc.cluster.local
    patch:
      operation: MERGE
      value:
        transport_socket_matches:
        - name: tlsMode-disabled
          transport_socket:
            name: envoy.transport_sockets.upstream_proxy_protocol
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.transport_sockets.proxy_protocol.v3.ProxyProtocolUpstreamTransport
              config:
                version: V1
              transport_socket:
                name: "envoy.transport_sockets.raw_buffer"
```

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go [hunk 1]
```diff
@@ -75,13 +75,61 @@ func mergeTransportSocketCluster(c *cluster.Cluster, cp *model.EnvoyFilterConfig
...
+	// First merge transport socket matches with the same name
+	if (len(c.GetTransportSocketMatches())) > 0 && (len(cpValueCast.GetTransportSocketMatches()) > 0) {
+		for _, cpTsm := range cpValueCast.GetTransportSocketMatches() {
+			if cpTsm.GetName() == "" || (cpTsm.GetTransportSocket() == nil && cpTsm.GetMatch() == nil) {
+				continue
+			}
+
+			for _, tsm := range c.GetTransportSocketMatches() {
+				if tsm.GetName() == cpTsm.GetName() {
+					// Replace match if defined in patch
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go:89]: I am not sure whether it is right

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go [hunk 2]
```diff
@@ -75,13 +75,61 @@ func mergeTransportSocketCluster(c *cluster.Cluster, cp *model.EnvoyFilterConfig
 		return false, fmt.Errorf("cast of cp.Value failed: %v", okCpCast)
 	}
 
+	merged = false
+
+	// First merge transport socket matches with the same name
+	if (len(c.GetTransportSocketMatches())) > 0 && (len(cpValueCast.GetTransportSocketMatches()) > 0) {
+		for _, cpTsm := range cpValueCast.GetTransportSocketMatches() {
+			if cpTsm.GetName() == "" || (cpTsm.GetTransportSocket() == nil && cpTsm.GetMatch() == nil) {
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go:83]: I can not understand this well, what do you do with Match check?

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go (hunk 1 above)
Comment (JuniorJPDJ) [reply pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go:89]: Is this about the condition or just about replacing whole match?

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go [hunk 3]
```diff
@@ -75,13 +75,61 @@ func mergeTransportSocketCluster(c *cluster.Cluster, cp *model.EnvoyFilterConfig
 		return false, fmt.Errorf("cast of cp.Value failed: %v", okCpCast)
 	}
 
+	merged = false
+
+	// First merge transport socket matches with the same name
+	if (len(c.GetTransportSocketMatches())) > 0 && (len(cpValueCast.GetTransportSocketMatches()) > 0) {
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go:81]: Can we merge this loop with L133

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go:83]: Why donot merge when both match = nil

Comment (JuniorJPDJ) [reply pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go:83]: As there are no other fields, so nothing is left to merge.

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go (hunk 3 above)
Comment (JuniorJPDJ) [reply pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go:81]: I can try, let me see.

Comment (JuniorJPDJ) [reply pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go:81]: It should be possible, but I would need to remove check in L127.
Doing it separately should avoid clutter and workarounds for making this integrated.

I added some comments to the whole process.

**Review by hzxuzhonghu — COMMENTED** [review]
Not to talk about tsm match merging itself, I am not sure this cluster merge is right, when we have one transportsocker match merged, we will skip all other cluster fields merging.

I think ideally we should introduce patch applyTo transport socket

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go [hunk 4]
```diff
@@ -75,13 +75,68 @@ func mergeTransportSocketCluster(c *cluster.Cluster, cp *model.EnvoyFilterConfig
...
+
+			for _, tsm := range c.GetTransportSocketMatches() {
+				if tsm.GetName() != cpTsm.GetName() {
+					// Skip unmatched names
+					continue
+				}
+
+				if cpTsm.GetMatch() != nil {
+					// Replace Match if defined in patch
+					tsm.Match = cpTsm.Match
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go:98]: NOt sure if this is right

**Comment by howardjohn** [comment]
I put a hold as I want to thoroughly review this since this code path caused multiple outages in the past and is tricky. However, I will not have time until next week due to KubeCon. I will review ASAP

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go (hunk 4 above)
Comment (JuniorJPDJ) [reply pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go:98]: We are not stopping merging on this, for is still running ;)

**Comment by JuniorJPDJ** [comment]
> Not to talk about tsm match merging itself, I am not sure this cluster merge is right, when we have one transportsocker match merged, we will skip all other cluster fields merging.

You are right actually, but this is the previous behaviour too.

> I think ideally we should introduce patch applyTo transport socket

You mean to specify `TransportSocket` instead of `Cluster` in `EnvoyFilter`? This would be cool, but I'll not do this, I'm not Golang programmer and I don't know the codebase ;)

**Comment by JuniorJPDJ** [comment]
I'll prepare commit for this refactor I mentioned above, we can revert it if it's not a good idea.

**Comment by JuniorJPDJ** [comment]
Ok, test passed.
Lets push refactor.

**Comment by JuniorJPDJ** [comment]
Ok, well...

```
?       istio.io/istio/pilot/pkg/security/authn/factory [no test files]                                                                                                                                                                                                                   
--- FAIL: TestClusterPatching (0.07s)                                                                                                                                                                                                                                                     
    --- FAIL: TestClusterPatching/sidecar_outbound_cluster_patch (0.03s)                                                                                                                                                                                                                  
        cluster_patch_test.go:698: sidecar outbound cluster patch mismatch (-want +got):                                                                                                                                                                                                  
              []*clusterv3.Cluster{                                                                                                                                                                                                                                                       
                Inverse(protocmp.Transform, protocmp.Message{"@type": s"envoy.config.cluster.v3.Cluster", "dns_lookup_family": s"V6_ONLY", "lb_policy": s"RING_HASH", "name": string("outbound|443||cluster1"), ...}),                                                                    
                Inverse(protocmp.Transform, protocmp.Message{                                                                                                                                                                                                                             
                        ... // 3 identical entries                                                                                                                                                                                                                                        
                        "lb_policy":        s"RING_HASH",                                                                                                                                                                                                                                 
                        "name":             string("outbound|443||cluster2"),                                                                                                                                                                                                             
            -           "transport_socket": s`{name:"envoy.transport_sockets.tls", typed_config:{[type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext]:{common_tls_context:{tls_params:{tls_minimum_protocol_version:TLSv1_3}}}}}`,                          
                }),                                                                                                                                                                                                                                                                       
                Inverse(protocmp.Transform, protocmp.Message{                                                                                                                                                                                                                             
                        ... // 2 identical entries                                                                                                                                                                                                                                        
                        "lb_policy": s"RING_HASH",                                                                                                                                                                                                                                        
                        "name":      string("outbound|443||cluster3"),                                                                                                                                                                                                                    
                        "transport_socket": protocmp.Message{                                                                                                                                                                                                                             
                                "@type": s"envoy.config.core.v3.TransportSocket",                                                                                                                                                                                                         
                                "name":  string("envoy.transport_sockets.tls"),                                                                                                                                                                                                           
                                "typed_config": protocmp.Message{                                                                                                                                                                                                                         
                                        "@type":    s"google.protobuf.Any",                                                                                                                                                                                                               
                                        "type_url": string("type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.Up"...),                                                                                                                                                        
                                        "value": protocmp.Message{                                                                                                                                                                                                                        
                                                "@type": s"envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext",                                                                                                                                                                 
                                                "common_tls_context": protocmp.Message{                                                                                                                                                                                                   
                                                        "@type":                                         s"envoy.extensions.transport_sockets.tls.v3.CommonTlsContext",                                                                                                                   
                                                        "tls_certificate_certificate_provider_instance": protocmp.Message{"@type": s"envoy.extensions.transport_sockets.tls.v3.CommonTlsContext.Certi"..., "certificate_name": string("certificate"), "instance_name": string("instance")}
,                                                                                                                                                                                                                                                                                         
                                                        "tls_params": protocmp.Message{                                                                                                                                                                                                   
                                                                "@type":                        s"envoy.extensions.transport_sockets.tls.v3.TlsParameters",                                                                                                                               
                                                                "ecdh_curves":                  []string{"X25519"},                                                                                                                                                                       
                                                                "tls_maximum_protocol_version": s"TLSv1_3",                                                                                                                                                                               
            -                                                   "tls_minimum_protocol_version": s"TLSv1_3",                                                                                                                                                                               
                                                        },                                                                                                                                                                                                                                
                                                },                                                                                                                                                                                                                                        
                                        },                                                                                                                                                                                                                                                
                                },                                                                                                                                                                                                                                                        
                        },                                                                                                                                                                                                                                                                
                }),                                                                                                                                                                                                                                                                       
                Inverse(protocmp.Transform, protocmp.Message{"@type": s"envoy.config.cluster.v3.Cluster", "dns_lookup_family": s"V6_ONLY", "lb_policy": s"RING_HASH", "name": string("outbound|7777||custom-tls-addition"), ...}),                                                        
                Inverse(protocmp.Transform, protocmp.Message{                                                                                                                                                                                                                             
                        ... // 2 identical entries                                                                                                                                                                                                                                        
                        "lb_policy": s"RING_HASH",                                                                                                                                                                                                                                        
                        "name":      string("outbound|7777||custom-tls-replacement"),                                                                                                                                                                                                     
                        "transport_socket": protocmp.Message{                                                                                                                                                                                                                             
                                "@type": s"envoy.config.core.v3.TransportSocket",                                                                                                                                                                                                         
            -                   "name":  string("transport_sockets.alts"),                                                                                                                                                                                                                
            +                   "name":  string("envoy.transport_sockets.tls"),                                                                                                                                                                                                           
                                "typed_config": protocmp.Message{                                                                                                                                                                                                                         
                                        "@type": s"google.protobuf.Any",                                                                                                                                                                                                                  
                                        "type_url": strings.Join({                                                                                                                                                                                                                        
                                                "type.googleapis.com/",                                                                                                                                                                                                                   
            -                                   "udpa.type.v1.TypedStruc",                                                                                                                                                                                                                
            +                                   "envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContex",                                                                                                                                                                            
                                                "t",                                                                                                                                                                                                                                      
                                        }, ""),                                                                                                                                                                                                                                           
                                        "value": protocmp.Message{                                                                                                                                                                                                                        
            -                                   "@type":              s"udpa.type.v1.TypedStruct",                                                                                                                                                                                        
            +                                   "@type":              s"envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext",                                                                                                                                                    
            +                                   "common_tls_context": s`{tls_params:{tls_minimum_protocol_version:TLSv1_1, ecdh_curves:["X25519"]}, tls_certificate_certificate_provider_instance:{instance_name:"instance", certificate_name:"certificate"}}`,                           
            -                                   "type_url":           string("type.googleapis.com/envoy.extensions.transport_sockets.alts.v3.Alts"),                                                                                                                                      
            -                                   "value":              s`{fields:{"handshaker_service":{string_value:"1.2.3.4"}}}`,                                                                                                                                                        
                                        },                                                                                                                                                                                                                                                
                                },                                                                                                                                                                                                                                                        
                        },                                                                                                                                                                                                                                                                
                }),                                                                                                                                                                                                                                                                       
                Inverse(protocmp.Transform, protocmp.Message{"@type": s"envoy.config.cluster.v3.Cluster", "dns_lookup_family": s"V6_ONLY", "lb_policy": s"RING_HASH", "name": string("outbound|7777||custom-tls-replacement-tsm"), ...}),                                                 
                Inverse(protocmp.Transform, protocmp.Message{"@type": s"envoy.config.cluster.v3.Cluster", "name": string("new-cluster1")}),                                                                                                                                               
                Inverse(protocmp.Transform, protocmp.Message{"@type": s"envoy.config.cluster.v3.Cluster", "name": string("new-cluster2")}),                                                                                                                                               
              }                                                                                                                                                                                                                                                                           
FAIL                                                                                                                                                                                                                                                                                      
FAIL    istio.io/istio/pilot/pkg/networking/core/v1alpha3/envoyfilter   0.303s     
```

**Comment by JuniorJPDJ** [comment]
I don't understand where it comes from.
I'll try to take care about it tomorrow.

**Comment by JuniorJPDJ** [comment]
Ok, previous version was storing modified version of patches, that's why I had problems with multiple clusters patch applying.
I created backup of `EnvoyFilterWrapper.Value` and I'm restoring the copy after merging. Now it works!

**Comment by JuniorJPDJ** [comment]
@hzxuzhonghu @howardjohn If you guys want I can make the latest commit separate MR as it changes more things, but for now I need some feedback ;)

**Comment by JuniorJPDJ** [comment]
Just rebased it.

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go [hunk 5]
```diff
@@ -50,38 +50,112 @@ func ApplyClusterMerge(pctx networking.EnvoyFilter_PatchContext, efw *model.Envo
...
+
+	if dstCluster != nil && srcPatch != nil {
+
+		retVal, errMerge := util.MergeAnyWithAny(dstCluster, srcPatch)
+		if errMerge != nil {
+			return fmt.Errorf("function MergeAnyWithAny failed for ApplyClusterMerge: %v", errMerge)
+		}
+
+		// Merge the above result with the whole cluster
+		merge.Merge(dstCluster, retVal)
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go:84]: hm, I saw we did this in another part of the code, but I don't get it.

Isn't retVal already `dstCluster + srcPatch`? So now we are doing `dstCluster = dstCluster + (dstCluster + srcPatch)` which seems like a NOP?

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go (hunk 4 above)
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go:121]: I guess if they explicitly set the match in the patch it seems ok? But usually they should not set the match probably.

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go [hunk 6]
```diff
@@ -50,38 +50,112 @@ func ApplyClusterMerge(pctx networking.EnvoyFilter_PatchContext, efw *model.Envo
...
+				}
+			}
+
+			if !tsmMerged {
+				newPatchTsms = append(newPatchTsms, cpTsm)
+			}
+		}
+
+		if len(newPatchTsms) != len(cpValueCast.TransportSocketMatches) {
+			cpValueCast.TransportSocket = nil
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go:147]: Not sure I understand why we are changing the patch value?

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go [hunk 7]
```diff
@@ -50,38 +50,112 @@ func ApplyClusterMerge(pctx networking.EnvoyFilter_PatchContext, efw *model.Envo
 			continue
 		}
 		if commonConditionMatch(pctx, cp) && clusterMatch(c, cp, hosts) {
+			// backup the Cluster in case mergeTransportSocketCluster modifies it
+			cpValueBkp := proto.Clone(cp.Value)
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go:54]: ideally we would only clone when we are actually mutating things, to avoid high cost for users not using this functionality

Comment (JuniorJPDJ) [reply pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go:54]: Patches aren't supposed to be terribly big, so this shouldn't be an issue. I suspect patching is not being called frequently too. How would you check if clone is needed? As I don't see simple solution and I think over-complicating already complicated code will probably not end well.

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go (hunk 5 above)
Comment (JuniorJPDJ) [reply pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go:84]: TBH. Me neither, so I decided to leave it as is. I suspect `util.MergeAnyWithAny` is doing some sort of preparation of result to replace and merge.Merge is doing a replacement.

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go (hunk 4 above)
Comment (JuniorJPDJ) [reply pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go:121]: This is replacing only if explicitly set in patch (nil check) and I already see some usecases for Match patching.
eg. When I want default TSM to handle part of the traffic, but I add another one after that which handles rest of it.

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go (hunk 6 above)
Comment (JuniorJPDJ) [reply pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go:147]: This is removing TransportSocket from patch if TSM patching occured, so we don't have TS in the patch when merging rest of patch later.
It is similar to the case I already found in the code - it is in line 169 now.

**Comment by JuniorJPDJ** [comment]
@howardjohn what are we going to do with this? I replied to all your concerns.

**Comment by klarose** [comment]
This seems to have stalled. What can we do to pick it back up again? I'm nervous about upgrading my cluster because of how fragile my workaround is.

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go (hunk 7 above)
Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go:54]: Patching is called frequently and most of the CPU time tends to be on cloning

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go:54]: In mergeTransportSocketCluster we are just overwriting TransportSocket or TransportSocketMatches, I think we can just shallowcopy it instead of deepcopy. This can be done in `protomarshal.ShallowCopy`, which should be much cheaper. I would also do it inside `mergeTransportSocketCluster`, so we don't to restore, I would make mergeTransportSocketCluster return the new patch and not modify the original cp.Value at all; modifying input params is always a little hard to follow and easy to make mistakes.

Comment (jpruciak) [reply pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go:54]: Seems like an idea. Let's try.

Comment (JuniorJPDJ) [reply pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go:54]: Seems like an idea. Let's try.

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go:103]: I understand you are checking tsm name, if names equal, can do tsm merging. not to care about match or ts fields here

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go [hunk 8]
```diff
@@ -50,38 +50,112 @@ func ApplyClusterMerge(pctx networking.EnvoyFilter_PatchContext, efw *model.Envo
...
+
+			if !tsmMerged {
+				newPatchTsms = append(newPatchTsms, cpTsm)
+			}
+		}
+
+		if len(newPatchTsms) != len(cpValueCast.TransportSocketMatches) {
+			cpValueCast.TransportSocket = nil
+		}
+		cpValueCast.TransportSocketMatches = newPatchTsms
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go:149]: will these be merged with cluster,actually append after

File: pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go [hunk 9]
```diff
@@ -50,38 +50,112 @@ func ApplyClusterMerge(pctx networking.EnvoyFilter_PatchContext, efw *model.Envo
...
 
 	// Check if cluster patch has a transport socket.
 	if cpValueCast.GetTransportSocket() == nil {
-		return false, nil
+		return nil
 	}
 	var tsmPatch *core.TransportSocket
 
-	// First check if the transport socket matches with any cluster transport socket matches.
+	// Check if the transport socket matches with any cluster transport socket matches.
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go:158]: not related to this patch, just to mark the below handling is very hacky.  I am even confused after reading carefully

**Comment by jpruciak** [comment]
can someone please reopen?

**Comment by hzxuzhonghu** [comment]
As i said ealier we may need a new TransportSocketMatch type patch

**Comment by linsun** [comment]
Hi @JuniorJPDJ, just checking if you are planning to address latest comments soon?

**Comment by linsun** [comment]
We probably need a release note for this as well.

**Comment by JuniorJPDJ** [comment]
Sure, but I lack the time now - it will need to wait for a bit more if noone else adresses the problems.
